src/kubernetes_cluster/spec/controller/controller_runtime.rs

```diff
@@ -48,52 +50,110 @@ pub open spec fn continue_reconcile<K: ResourceView, R: Reconciler<K>>() -> Cont
                 &&& s.ongoing_reconciles.dom().contains(cr_key)
                 &&& !R::reconcile_done(s.ongoing_reconciles[cr_key].local_state)
                 &&& !R::reconcile_error(s.ongoing_reconciles[cr_key].local_state)
-                &&& if input.recv.is_Some() {
+                // Split cases:
+                // (1) there is a pending request which is destined for kubernetes api;
+                // (2) there is a pending request which is destined for external library and no response;
+                // (3) there is a pending request which is destined for external library and a corresponding response;
+                // (4) there is no pending request;
```

Comment:
Since we don't plan to have a separate state machine for external lib, I am wondering whether we want to split (2) and (3). An alternative is that in the same action, if the controller issues an external lib call, we run the "process" function and get the return value of the external lib and then store it in `lib_response`. Of course, this model has some limitations in its expressiveness, but it is better to start from a very simple one.